Mathlib/AlgebraicGeometry/Noetherian.lean

```diff
@@ -0,0 +1,446 @@
+If the localization of `R` at each `f i` is noetherian, so is `R`.
+
+We follow the proof given in [Har77], Proposition II.3.2 -/
+theorem isNoetherianRing_of_away : IsNoetherianRing R := by
+  apply Iff.mp
+  apply monotone_stabilizes_iff_noetherian
+  intro I
+  let floc s := algebraMap R (Away (M := R) s)
+  let suitableN s :=
+    { n : ℕ | ∀ m : ℕ, n ≤ m → (Ideal.map (floc s) (I n)) = (Ideal.map (floc s) (I m)) }
+  let minN s := sInf (suitableN s)
+  have hSuit : ∀ s : S, minN s ∈ suitableN s := by
+    intro s
+    apply Nat.sInf_mem
+    let f : ℕ →o Ideal (Away (M := R) s) :=
+      ⟨fun n => Ideal.map (floc s) (I n), by
+        intro n m hnm
+        dsimp
+        apply Ideal.map_mono
+        exact I.monotone hnm⟩
+    exact monotone_stabilizes_iff_noetherian.mpr (hN s) f
+  let N := Finset.sup S minN
+  use N
+  have hN : ∀ s : S, minN s ≤ N := fun s => Finset.le_sup s.prop
+  intro n hn
+  rw [IsLocalization.ideal_eq_iInf_comap_map_away hS (I N),
+      IsLocalization.ideal_eq_iInf_comap_map_away hS (I n),
+      iInf_subtype', iInf_subtype']
+  apply iInf_congr
+  intro s
+  congr 1
+  rw [← hSuit s N (hN s)]
+  exact hSuit s n <| Nat.le_trans (hN s) hn
+
+end localizationProps
+
+variable {X : Scheme}
+
+/-- If a scheme `X` has a cover by affine opens whose sections are Noetherian rings,
+then `X` is locally Noetherian. -/
+theorem isLocallyNoetherian_of_affine_cover (S : Set X.affineOpens)
+    (hS : (⨆ i : S, i : Opens X) = ⊤)
+    (hS' : ∀ (U : S), IsNoetherianRing (X.presheaf.obj (op U))) : IsLocallyNoetherian X := by
+  refine ⟨fun U => ?_⟩
+  have hS : ⋃ i : S, (i : Set X) = Set.univ := by
+    rw [← (Opens.coe_iSup _)]
+    apply Opens.coe_eq_univ.mpr
+    exact hS
+  apply of_affine_open_cover (P := _) U S _ _ hS hS'
+  · intro U f hN
+    let R := X.presheaf.obj (op U)
+    let Rf := Localization.Away f
+    have hh : IsLocalization _ Rf := isLocalization
+    have : IsNoetherianRing Rf := by
+      apply @IsLocalization.isNoetherianRing R _ _ Rf _ algebra hh
+      assumption
+    let Rf' := X.presheaf.obj (op (X.basicOpen f))
+    have hAff := IsAffineOpen.isLocalization_basicOpen U.prop f
+    have := @IsLocalization.algEquiv R _ _ Rf _ _ hh Rf' _ _ hAff
+    apply isNoetherianRing_of_ringEquiv Rf
+    exact this.toRingEquiv
+  · intro U s _ hN
+    let R := X.presheaf.obj (op U)
+    have : ∀ f : s, IsNoetherianRing (Away (M := R) f) := by
+      intro ⟨f, hf⟩
+      have : IsNoetherianRing (X.presheaf.obj (op (X.basicOpen f))) := hN ⟨f, hf⟩
+      apply isNoetherianRing_of_ringEquiv (X.presheaf.obj (op (X.basicOpen f)))
+      let Rf := Localization.Away f
+      have hh : IsLocalization _ Rf := isLocalization
+      let Rf' := X.presheaf.obj (op (X.basicOpen f))
+      have hAff := IsAffineOpen.isLocalization_basicOpen U.prop f
+      have := @IsLocalization.algEquiv R _ _ Rf _ _ hh Rf' _ _ hAff
+      exact this.symm.toRingEquiv
```

Comment:
Similarly this can be shortened into 
```
      have : IsNoetherianRing Γ(X, X.basicOpen f) := hN ⟨f, hf⟩
      apply isNoetherianRing_of_ringEquiv Γ(X, X.basicOpen f)
      have := IsLocalization.algEquiv (.powers f) (Localization.Away f) Γ(X, X.basicOpen f)
      exact this.symm.toRingEquiv
```